### PR TITLE
Improve identity provider and mapper handling

### PR DESCRIPTION
**Pull request by spahrson**
I added support for updating only changed IdentityProviders and the possibility to use 'managed' IdentityProvider. Furthermore I added IdentityProviderMappers (create, update, delete, managed).

### CODE REVIEW

**Comment by jkroepke** [comment]
All changed java files have indent 2 instead indent 4.

**Comment by spahrson** [comment]
> All changed java files have indent 2 instead indent 4.
> 
> Please use indent 4.

I updated the editor config too.

File: src/main/java/de/adorsys/keycloak/config/properties/ImportConfigProperties.java [hunk 1]
```diff
@@ -119,18 +119,27 @@ public boolean isParallel() {
         @NotNull
         private final ImportManagedPropertiesValues authenticationFlow;
 
+        @NotNull
+        private final ImportManagedPropertiesValues identityProvider;
```
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/properties/ImportConfigProperties.java:123]: Please add a test in `ImportConfigPropertiesTest` for new properties.

File: src/main/java/de/adorsys/keycloak/config/repository/IdentityProviderMapperRepository.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        return loadIdentityProviderMapperByName(realm, identityProviderAlias, name);
+    }
+
+    public IdentityProviderMapperRepresentation getIdentityProviderMapperByName(String realm, String identityProviderAlias, String name) {
+        Optional<IdentityProviderMapperRepresentation> maybeIdentityProviderMapper = loadIdentityProviderMapperByName(realm, identityProviderAlias, name);
+
+        if (maybeIdentityProviderMapper.isPresent()) {
+            return maybeIdentityProviderMapper.get();
+        }
+        return null;
```
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/repository/IdentityProviderMapperRepository.java:58]:
```suggestion
        return maybeIdentityProviderMapper.orElse(null);
```

File: src/main/java/de/adorsys/keycloak/config/service/IdentityProviderImportService.java [hunk 3]
```diff
@@ -21,54 +21,161 @@
...
+        IdentityProviderMapperRepresentation existingIdentityProviderMapper = identityProviderMapperRepository.getIdentityProviderMapperByName(realm, identityProviderMapper.getIdentityProviderAlias(), identityProviderMapper.getName());
+        IdentityProviderMapperRepresentation patchedIdentityProviderMapper = CloneUtil.patch(existingIdentityProviderMapper, identityProviderMapper);
+        String identityProviderMapperName = existingIdentityProviderMapper.getName();
+        String identityProviderAlias = existingIdentityProviderMapper.getIdentityProviderAlias();
+
+        if (isIdentityProviderMapperEqual(existingIdentityProviderMapper, patchedIdentityProviderMapper)) {
+            logger.debug("No need to update identityProviderMapper for identityProvider '{}' in realm '{}' in realm '{}'", identityProviderMapperName, identityProviderAlias, realm);
+        } else {
+            logger.debug("Update identityProviderMapper '{}' for identityProvider '{}' in realm '{}'", identityProviderMapperName, identityProviderAlias, realm);
+            identityProviderMapperRepository.updateIdentityProviderMapper(realm, patchedIdentityProviderMapper);
```
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/service/IdentityProviderImportService.java:159]: Could you write a test which covers this lines?

**Comment by jkroepke** [comment]
> to use 'managed'

Off-topic: Do you have a better name for this?

**Comment by spahrson** [comment]
> > to use 'managed'
> 
> Off-topic: Do you have a better name for this?

'managed' is ok for me. but i had to read your doc to understand what affects this property has. it's kind of synchronization control.

File: src/test/java/de/adorsys/keycloak/config/test/util/KeycloakVersion.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    private static final String KEYCLOAK_VERSION_ENV_NAME = "KEYCLOAK_VERSION";
+
+    public static boolean isKeycloak8() {
+        String version = getKeycloakVersion();
+        return version != null && version.startsWith("8.");
+    }
+
+    private static String getKeycloakVersion() {
+        return System.getenv(KEYCLOAK_VERSION_ENV_NAME);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/test/util/KeycloakVersion.java:33]: Could you use `System.getProperty("keycloak.version");` instead `System.getenv` for consistency?

Like: 
https://github.com/adorsys/keycloak-config-cli/blob/f7ed19ab9a8b7e4f2191317cba436b63f20802f7/src/test/java/de/adorsys/keycloak/config/ImportExportedRealmIT.java#L45

It uses the value from maven directly, e.g. this is not covering `mvn clean verify -Dkeycloak.version=8.0.2`

File: src/test/java/de/adorsys/keycloak/config/test/util/KeycloakVersion.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    private static final String KEYCLOAK_VERSION_ENV_NAME = "KEYCLOAK_VERSION";
+
+    public static boolean isKeycloak8() {
+        String version = getKeycloakVersion();
+        return version != null && version.startsWith("8.");
+    }
+
+    private static String getKeycloakVersion() {
+        return System.getProperty(KEYCLOAK_VERSION_ENV_NAME);
+    }
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/test/util/KeycloakVersion.java:34]:
```suggestion
    private static final String KEYCLOAK_VERSION_PROPERTY_NAME = "keycloak.version";

    public static boolean isKeycloak8() {
        String version = getKeycloakVersion();
        return version != null && version.startsWith("8.");
    }

    private static String getKeycloakVersion() {
        return System.getProperty(KEYCLOAK_VERSION_PROPERTY_NAME);
    }
```

**Pull request merged**